### PR TITLE
Merge rabbitmq/cuttlefish changes.

### PR DESCRIPTION
**Pull request by hairyhum**
Addresses https://github.com/rabbitmq/cuttlefish/issues/3

### CODE REVIEW

**Comment by Licenser** [comment]
cc @tsloughter

File: rebar.config [hunk 1]
```diff
@@ -1,20 +1,130 @@
...
+%% under the License.
+%%
+%% -------------------------------------------------------------------
 {minimum_otp_vsn, "R16"}.
 
-{erl_opts, [warnings_as_errors,
-            {parse_transform, lager_transform},
-            debug_info,
-            warn_untyped_record]}.
+{brt_protect, true}.
```
Comment (Licenser) [inline rebar.config:23]: what's this doing?

File: rebar.config [hunk 2]
```diff
@@ -1,20 +1,130 @@
...
-{escript_emu_args, "%%! -escript main cuttlefish_escript +S 1 +A 0\n"}.
 {escript_incl_apps, [goldrush, getopt, lager, cuttlefish]}.
 {escript_main_app, cuttlefish}.
 
-{provider_hooks, [{post, [{compile, {default, escriptize}}]}]}.
+{provider_hooks, [
+    {pre,   [{compile,  neotoma}]}
+]}.
+
+{deps, [
```
Comment (Licenser) [inline rebar.config:45]: dependencies need to be hex dependencies

File: rebar.config [hunk 3]
```diff
@@ -1,20 +1,130 @@
...
-{deps, [getopt, lager]}.
+{escript_emu_args,
+    "%%! -escript main cuttlefish_escript +S 1 +A 0\n"}.
 
-{escript_emu_args, "%%! -escript main cuttlefish_escript +S 1 +A 0\n"}.
 {escript_incl_apps, [goldrush, getopt, lager, cuttlefish]}.
 {escript_main_app, cuttlefish}.
 
-{provider_hooks, [{post, [{compile, {default, escriptize}}]}]}.
+{provider_hooks, [
```
Comment (Licenser) [inline rebar.config:41]: the current solution to use neotoma only in the dev profilse seems better then hard including it

File: rebar.config [hunk 4]
```diff
@@ -1,20 +1,130 @@
...
+        {erl_opts, [
+            strong_validation,
+            warn_missing_spec,
+            warn_untyped_record,
+            {d, 'BASHO_VALIDATE'}
+        ]}
+    ]}
+]}.
+
+{plugins, [
```
Comment (Licenser) [inline rebar.config:117]: what are the basho rebar tools? Where are they used that wasn't required before?

File: Makefile [hunk 5]
```diff
@@ -0,0 +1,111 @@
+# -------------------------------------------------------------------
```
Comment (Licenser) [inline Makefile:1]: Do we really need a makefile?

File: src/cuttlefish.app.src [hunk 6]
```diff
@@ -1,15 +1,11 @@
...
+{application, cuttlefish, [
+    {description,   "cuttlefish configuration abstraction"},
+    {vsn,           git},
+    {registered,    []},
+    {applications,  [kernel, sasl, stdlib, syntax_tools, getopt, lager]},
+    {env,           []},
+    {modules,       []},
+    {maintainers,   ["RabbitMQ Team @ Pivotal Software"]},
+    {licenses,      ["Apache 2.0"]},
+    {links,         [{"GitHub",  "https://github.com/rabbitmq/cuttlefish"}]}
```
Comment (Licenser) [inline src/cuttlefish.app.src:10]: was wrong before too but that I look at it now this should probably point to Kyorai

File: src/cuttlefish_rebar_plugin.erl [hunk 7]
```diff
@@ -21,75 +18,120 @@
...
+%% For now, this is specific to Rebar2, but it's been refactored to allow it
+%% to be extended to work with Rebar3 as well.
+%% When we decide how we want to use it, we should do that.
+%%
 -module(cuttlefish_rebar_plugin).
 
 -export([
-    generate/2
+    generate/2,
+    init/1
```
Comment (Licenser) [inline src/cuttlefish_rebar_plugin.erl:33]: I think there is a seperate rebar3 cuttlefish plugin by @tsloughter (https://github.com/tsloughter/rebar3_cuttlefish) any reason for having a plugin in both here and a separate repo?

**Review by Licenser — CHANGES_REQUESTED** [review]
I stopped reading at cuttlefish_unit because I've the feeling :develop is the right branch @tsloughter can you give input on that?

**Comment by lukebakken** [comment]
@Licenser as you can see there are too many changes here, sorry about that. We'll more carefully comb through differences in `develop`.

**Comment by Licenser** [comment]
Oh that wasn't the issue (I'm on the same boat w/ _core a gazzilion changes) it was just kind of worried that we merge against the wrong branch.

**Comment by lukebakken** [comment]
No problem! It'll be easier to review several small PRs anyway 😄

**Pull request closed without merging**